pkg/api/api.go

```diff
@@ -0,0 +1,676 @@
+		RealtimeURL:          "https://realtime.enterprise.wikimedia.com/",
+	}
+
+	for _, opt := range ops {
+		opt(clt)
+	}
+
+	return clt
+}
+
+// Client is a struct that represents an HTTP client used to interact with the API.
+type Client struct {
+	// HTTPClient is the HTTP client used to send requests.
+	HTTPClient *http.Client
+
+	// UserAgent is the user-agent header value sent with each request.
+	UserAgent string
+
+	// BaseUrl is the base URL for all API requests.
+	BaseUrl string
+
+	// RealtimeURL is the base URL for all realtime API requests.
+	RealtimeURL string
+
+	// AccessToken is the access token used to authenticate requests.
+	AccessToken string
+
+	// DownloadMinChunkSize is the minimum chunk size used for downloading resources.
+	DownloadMinChunkSize int
+
+	// DownloadChunkSize is the chunk size used for downloading resources.
+	DownloadChunkSize int
+
+	// DownloadConcurrency is the number of simultaneous downloads allowed.
+	DownloadConcurrency int
+}
+
+func (c *Client) newRequest(ctx context.Context, url string, mtd string, pth string, req *Request) (*http.Request, error) {
+	dta := []byte{}
+
+	if req != nil {
+		bdy, err := json.Marshal(req)
+
+		if err != nil {
+			return nil, err
+		}
+
+		dta = bdy
+	}
+
+	hrq, err := http.NewRequestWithContext(ctx, mtd, fmt.Sprintf("%sv2/%s", url, pth), bytes.NewReader(dta))
+
+	if err != nil {
+		return nil, err
+	}
+
+	hrq.Header.Set("User-Agent", c.UserAgent)
+	hrq.Header.Set("Content-Type", "application/json")
+	hrq.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.AccessToken))
+
+	return hrq, nil
+}
+
+func (c *Client) do(hrq *http.Request) (*http.Response, error) {
+	res, err := c.HTTPClient.Do(hrq)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if res.StatusCode < http.StatusOK || res.StatusCode > http.StatusIMUsed {
+		dta, err := io.ReadAll(res.Body)
+		defer res.Body.Close()
+
+		if err != nil {
+			return nil, err
+		}
+
+		if len(string(dta)) == 0 {
+			return nil, errors.New(res.Status)
+		}
+
+		return nil, errors.New(string(dta))
+	}
+
+	return res, nil
+}
+
+func (c *Client) getEntity(ctx context.Context, req *Request, pth string, val interface{}) error {
+	hrq, err := c.newRequest(ctx, c.BaseUrl, http.MethodPost, pth, req)
+
+	if err != nil {
+		return err
+	}
+
+	res, err := c.do(hrq)
+
+	if err != nil {
+		return err
+	}
+
+	defer res.Body.Close()
+	return json.NewDecoder(res.Body).Decode(val)
+}
+
+func (c *Client) readAll(ctx context.Context, rdr io.Reader, cbk ReadCallback) error {
+	gzr, err := pgzip.NewReader(rdr)
+
+	if err != nil {
+		return err
+	}
+
+	trr := tar.NewReader(gzr)
+
+	for {
+		_, err := trr.Next()
+
+		if err == io.EOF {
+			break
+		}
+
+		if err != nil {
+			return err
+		}
+
+		if err := c.readLoop(ctx, trr, cbk); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *Client) readLoop(ctx context.Context, rdr io.Reader, cbk ReadCallback) error {
+	scn := bufio.NewScanner(rdr)
+	scn.Buffer([]byte{}, 20971520)
```

Comment:
Should we define this up top as constant?

Reply:
Good point, I moved it to the client configuration instead of the constant to keep everything consistent.